### PR TITLE
[mlir] [memref] add more checks to the memref.reinterpret_cast

### PR DESCRIPTION
**Pull request by cxy-1993**
Operation memref.reinterpret_cast was accept input like:

  %out = memref.reinterpret_cast %in to offset: [%offset], sizes: [10], strides: [1]
         : memref<?xf32> to memref<10xf32>

A problem arises: while lowering, the true offset of %out is %offset, but its data type indicates an offset of 0. Permitting this inconsistency can result in incorrect outcomes, as certain pass might erroneously extract the offset from the data type of %out.

 This patch fixes this by enforcing that the return value's data type aligns
with the input parameter.

### CODE REVIEW

**Review by matthias-springer — APPROVED** [review]

File: mlir/test/Dialect/MemRef/invalid.mlir [hunk 1]
```diff
@@ -217,6 +217,15 @@ func.func @memref_reinterpret_cast_no_map_but_offset(%in: memref<?xf32>) {
 
 // -----
 
+func.func @memref_reinterpret_cast_offset_mismatch_dynamic(%in: memref<?xf32>, %offset : index) {
+  // expected-error @+1 {{expected result type with offset = -9223372036854775808 instead of 0}}
```
Comment (MaheshRavishankar) [inline mlir/test/Dialect/MemRef/invalid.mlir:221]: THis error message is hard to digest, and is a leak of the sentinal value used to represent dynamic shape. We need a better error message here.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: mlir/test/Dialect/MemRef/invalid.mlir (hunk 1 above)
Comment (cxy-1993) [reply mlir/test/Dialect/MemRef/invalid.mlir]: done

**Comment by cxy-1993** [comment]
Hi @MaheshRavishankar ,do you have any further comments on this patch?

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp [hunk 2]
```diff
@@ -1892,11 +1892,12 @@ LogicalResult ReinterpretCastOp::verify() {
   // Match sizes in result memref type and in static_sizes attribute.
   for (auto [idx, resultSize, expectedSize] :
        llvm::enumerate(resultType.getShape(), getStaticSizes())) {
-    if (!ShapedType::isDynamic(resultSize) &&
-        !ShapedType::isDynamic(expectedSize) && resultSize != expectedSize)
+    if (!ShapedType::isDynamic(resultSize) && resultSize != expectedSize)
       return emitError("expected result type with size = ")
-             << expectedSize << " instead of " << resultSize
-             << " in dim = " << idx;
+             << (ShapedType::isDynamic(expectedSize)
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1897]: If I am reading the logic correctly, this check is not needed (here and everywhere below.)

**Review by MaheshRavishankar — APPROVED** [review]
Tried to look through all the lit test changes. They all look good to me visually.

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp (hunk 2 above)
Comment (cxy-1993) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:1897]: Thank you for your review. However, I believe these checks are necessary. As the PR description mentions, mismatches between return value data types and operands may lead to other transforms incorrectly obtaining values of the wrong data type, resulting in erroneous outcomes. This submission is specifically designed to address this issue.

**Pull request merged**